sharding/contracts/validator_manager.sol

```diff
@@ -0,0 +1,248 @@
+    return validators[
+      int(
+        uint(
+          keccak256(
+            uint(
+              block.blockhash(_period - lookAheadPeriods)
+            ) * periodLength,
+            _shardId
+          )
+        ) %
+        uint(getValidatorsMaxIndex())
+      )
+    ].addr;
+  }
+
+  function deposit() public payable returns(int) {
+    require(!isValidatorDeposited[msg.sender]);
+    require(msg.value == depositSize);
+    // Find the empty slot index in validators set
+    int index;
+    if (!isStackEmpty())
+      index = stackPop();
+    else
+      index = int(numValidators);
+      
+    validators[index] = Validator({
+      deposit: msg.value,
+      addr: msg.sender
+    });
+    ++numValidators;
+    isValidatorDeposited[msg.sender] = true;
+    
+    Deposit(msg.sender, index);
+    return index;
+  }
+
+  // Removes the validator from the validator set and refunds the deposited ether 
+  function withdraw(int _validatorIndex) public {
+    require(msg.sender == validators[_validatorIndex].addr);
+    // [FIXME] Should consider calling the validator's contract, might be useful
+    // when the validator is a contract.
+    validators[_validatorIndex].addr.transfer(validators[_validatorIndex].deposit);
+    isValidatorDeposited[validators[_validatorIndex].addr] = false;
+    delete validators[_validatorIndex];
+    stackPush(_validatorIndex);
+    --numValidators;
+    Withdraw(_validatorIndex);
+  }
+
+  // Attempts to process a collation header, returns true on success, reverts on failure.
+  function addHeader(int _shardId, uint _expectedPeriodNumber, bytes32 _periodStartPrevHash,
+                     bytes32 _parentCollationHash, bytes32 _txListRoot,
+                     address _collationCoinbase, bytes32 _postStateRoot, bytes32 _receiptRoot,
+                     int _collationNumber) public returns(bool) {
+    HeaderVars memory headerVars;
+
+    // Check if the header is valid
+    require((_shardId >= 0) && (_shardId < shardCount));
+    require(block.number >= periodLength);
+    require(_expectedPeriodNumber == block.number / periodLength);
+    require(_periodStartPrevHash == block.blockhash(_expectedPeriodNumber * periodLength - 1));
+
+    // Check if this header already exists
+    headerVars.entireHeaderHash = keccak256(_shardId, _expectedPeriodNumber, _periodStartPrevHash,
+                                   _parentCollationHash, _txListRoot, bytes32(_collationCoinbase),
+                                   _postStateRoot, _receiptRoot, _collationNumber);
+    assert(headerVars.entireHeaderHash != 0x0);
+    assert(collationHeaders[_shardId][headerVars.entireHeaderHash].score == 0);
+    // Check whether the parent exists.
+    // if (parent_collation_hash == 0), i.e., is the genesis,
+    // then there is no need to check.
+    if (_parentCollationHash != 0x0)
+        assert((_parentCollationHash == 0x0) || (collationHeaders[_shardId][_parentCollationHash].score > 0));
```

Comment:
_parentCollationHash == 0x0 can never be true from within this if statement, please update